finality-aleph/src/data_io.rs

```diff
@@ -203,138 +448,535 @@ where
+                    TopBlockImportedButNotFinalizedAncestor
+                }
+            }
+            TopBlockImportedButIncorrectBranch => {
+                // This could also be a hopeless fork, but we have checked before that it isn't (yet).
+                TopBlockImportedButIncorrectBranch
+            }
+            AvailableHopelessFork | AvailableBlock => {
+                error!(target: "afa", "Checking availability of an available proposal {:?} status: {:?}", proposal, old_status);
+                old_status.clone()
+            }
+        }
+    }
+
+    // For a proposal that might be new or not, check if it is available. If it is a new proposal
+    // and it is not available:
+    // 1) create an entry in pending_proposals with an empty list of pending messages,
+    // 2) register an appropriate event trigger (for block import and/or finality).
+    fn is_proposal_available_prior_to_adding(&mut self, proposal: &AlephProposal<B>) -> bool {
+        if self.available_proposal_cache.contains(proposal) {
+            return true;
+        }
+        if self.pending_proposals.contains_key(proposal) {
+            return false;
+        }
+        use ProposalStatus::*;
+        let status = self.check_proposal_availability(proposal, &PendingTopBlock, true);
+        match status {
+            PendingTopBlock => {
+                self.pending_proposals
+                    .insert(proposal.clone(), PendingProposalInfo::new(status));
+                self.register_block_import_trigger(proposal, &proposal.top_block());
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            TopBlockImportedButIncorrectBranch => {
+                self.pending_proposals
+                    .insert(proposal.clone(), PendingProposalInfo::new(status));
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            TopBlockImportedButNotFinalizedAncestor => {
+                self.pending_proposals
+                    .insert(proposal.clone(), PendingProposalInfo::new(status));
+                self.register_next_finality_trigger(proposal);
+                false
+            }
+            AvailableBlock | AvailableHopelessFork => {
+                self.available_proposal_cache.put(proposal.clone(), ());
                 true
-            })
-            .collect();
+            }
+        }
+    }
+
+    fn on_message_dependencies_resolved(&self, message: Message) {
+        trace!(target: "afa", "Sending message from DataStore {:?}", message);
+        if let Err(e) = self.messages_for_aleph.unbounded_send(message) {
+            error!(target: "afa", "Unable to send a ready message from DataStore {}", e);
+        }
+    }
+
+    fn assign_fresh_message_id(&mut self) -> MessageId {
+        self.next_free_id += 1;
+        self.next_free_id - 1
```

Comment:
:vomiting_face: where is the post-increment operator in Rust?!!1?!

Reply:
Yeah, I didn't know how to do that better actually :)